### PR TITLE
OSL validation test setup

### PR DESCRIPTION
**Pull request by bernardkwok**
Add usage of external oslc to test validity of generated osl code.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -17,6 +17,10 @@ option(MATERIALX_INSTALL_PYTHON "Install the MaterialX Python package as a third
 option(MATERIALX_WARNINGS_AS_ERRORS "Interpret all compiler warnings as errors." OFF)
 option(MATERIALX_BUILD_VIEW "Build the MaterialX View module. " ON)
 option(MATERIALX_TEST_VIEW "Run tests for MaterialX View module. Requires a valid GPU setup." OFF)
+set(MATERIALX_OSLC_EXECUTABLE "" CACHE FILEPATH
```
Comment (bernardkwok) [inline CMakeLists.txt]: Make these build options.

File: source/MaterialXTest/ShaderGen.cpp [hunk 2]
```diff
@@ -960,6 +961,45 @@ TEST_CASE("Swizzling", "[shadergen]")
     REQUIRE(test2.getSourceCode() == test2Result);
 }
 
+//
+// Utility to call validate OSL. 
+// For now only call into oslc to compile an OSL file and get the results.
+//
+static void validateOSL(const std::string oslFileName, std::string& errorResult)
```
Comment (bernardkwok) [inline source/MaterialXTest/ShaderGen.cpp]: Use stand-alone function for now w/o validator.

File: source/MaterialXView/ShaderValidators/Osl/OslValidator.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+OslValidator::~OslValidator()
+{
+}
+
+void OslValidator::initialize()
+{
+}
+
+void OslValidator::compileOSL(const std::string oslFileName)
```
Comment (bernardkwok) [inline source/MaterialXView/ShaderValidators/Osl/OslValidator.cpp]: utility method on OSL validator class.

File: source/MaterialXView/ShaderValidators/Osl/OslValidator.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    void save(const std::string& fileName) override;
+    
+    /// @}
+    /// @name Compilation settings
+    /// @{
+
+    /// Set OSL executable path string. Note that it is currently assumed that this
+    /// references the location of the oslc executable.
+    /// @param executable Full path to OSL compiler executable
+    void setOslCompilerExecutable(const std::string executable)
```
Comment (bernardkwok) [inline source/MaterialXView/ShaderValidators/Osl/OslValidator.h]: User needs to set compiler path + include paths when using the osl validator.

File: source/MaterialXTest/ShaderGen.cpp (hunk 2 above)
Comment (niklasharrysson) [inline source/MaterialXTest/ShaderGen.cpp]: Missing reference & on oslFileName

File: source/MaterialXTest/ShaderGen.cpp [hunk 5]
```diff
@@ -2354,21 +2464,30 @@ TEST_CASE("Osl Output Types", "[shadergen]")
...
-        file.open(exampleName + "_vector2.osl");
+        std::string fileName(exampleName + "_vector2.osl");
+        file.open(fileName);
         file << shader->getSourceCode();
         file.close();
 
+        // TODO: Use validation in MaterialXView library
+        std::string errorResult;
+        validateOSL(fileName, errorResult);
+        //REQUIRE(errorResult.size() == 0);
```
Comment (niklasharrysson) [inline source/MaterialXTest/ShaderGen.cpp]: Some of the tests have been disabled like this. Are these compilations failing now? Maybe add a TODO to fix the errors and re-enable the tests later.

File: source/MaterialXView/ShaderValidators/Osl/OslValidator.h [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+#include <MaterialXView/ShaderValidators/ShaderValidator.h>
+#include <MaterialXView/ShaderValidators/ExceptionShaderValidationError.h>
+#include <MaterialXView/Handlers/ImageHandler.h>
+#include <vector>
+#include <string>
+#include <map>
+
+namespace MaterialX
+{
+// Shared pointer to a OslValidator
```
Comment (niklasharrysson) [inline source/MaterialXView/ShaderValidators/Osl/OslValidator.h]: Typo: a -> an

File: source/MaterialXView/ShaderValidators/Osl/OslValidator.h [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+/// The main services provided are:
+///     - Validation: Via oslc 
+///     - Introspection: None at this time.
+///     - Binding: None at this time.
+///     - Rendering: None at this time.
+///
+class OslValidator : public ShaderValidator
+{
+  public:
+    /// Create a GLSL validator instance
```
Comment (niklasharrysson) [inline source/MaterialXView/ShaderValidators/Osl/OslValidator.h]: GLSL -> OSL

File: source/MaterialXTest/ShaderGen.cpp (hunk 5 above)
Comment (bernardkwok) [reply source/MaterialXTest/ShaderGen.cpp]: Actually, I think the majority of the tests fail for now. I'll leave a general TODO.

File: source/MaterialXView/ShaderValidators/Osl/OslValidator.h [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    /// @}
+    /// @name Validation
+    /// @{
+
+    /// Validate creation of program based on an input shader
+    /// @param shader Input shader
+    void validateCreation(const ShaderPtr shader) override;
+
+    /// Validate creation of program based input shader stage strings
+    /// @param shader Input stages List of stage string
```
Comment (niklasharrysson) [inline source/MaterialXView/ShaderValidators/Osl/OslValidator.h]: This docstring needs a bit of rewording.

**Comment by niklasharrysson** [comment]
Looks good and I only had some minor comments above.

**Pull request merged**